README.md
[![awscdk-jsii-template](https://img.shields.io/badge/built%20with-awscdk--jsii--template-blue)](https://github.com/pahud/awscdk-jsii-template)
[![NPM version](https://badge.fury.io/js/cdk-spot-one.svg)](https://badge.fury.io/js/cdk-spot-one)
[![PyPI version](https://badge.fury.io/py/cdk-spot-one.svg)](https://badge.fury.io/py/cdk-spot-one)
![Release](https://github.com/pahud/cdk-spot-one/workflows/Release/badge.svg)

# cdk-spot-one

One spot instance with EIP and defined duration. No interruption.

# Why

Sometimes we need an Amazon EC2 instance with static fixed IP for testing or development purpose for a duration of
time(probably hours). We need to make sure during this time, no interruption will occur and we don't want to pay
for on-demand rate. `cdk-spot-one` helps you reserve one single spot instance with pre-allocated or new
Elastic IP addresses(EIP) with defined `blockDuration`, during which time the spot instance will be secured with no spot interruption.

Behind the scene, `cdk-spot-one` provisions a spot fleet with capacity of single instance for you and it associates the EIP with this instance. The spot fleet is reserved as spot block with `blockDuration` from one hour up to six hours to ensure the high availability for your spot instance.

Multiple spot instances are possible by simply specifying the `targetCapacity` construct property, but we only associate the EIP with the first spot instance at this moment.

Enjoy your highly durable one spot instance with AWS CDK!

# Constructs

This library provides two major constructs:

## SpotInstance

- Create a spot instance **without** any fleet
- Does **NOT** support [Spot Block](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances)
- Support `stop` or `hibernate` instance

Scenario: To leverage the `stop` or `hibernate` capabilities of the spot instance to persist the data in the ebs volume.


## SpotFleet

- Create a spot instance with a [Spot Fleet](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet.html)
- Support [Spot Block](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances)
- Does **NOT** support `stop` or `hibernate` instance

Scenario: To ensure the availability with no disruption with defined period up to 6 hours.

# Sample

## SpotInstance

```ts
import { SpotInstance } from 'cdk-spot-one';

new SpotInstance(stack, 'SpotInstance');
```

## SpotFleet

```ts
import { SpotFleet } from 'cdk-spot-one';

// create the first fleet for one hour and associate with our existing EIP
const fleet = new SpotFleet(stack, 'SpotFleet')

// configure the expiration after 1 hour
fleet.expireAfter(Duration.hours(1))


// create the 2nd fleet with single Gravition 2 instance for 6 hours and associate with new EIP
const fleet2 = new SpotFleet(stack, 'SpotFleet2', {
  blockDuration: BlockDuration.SIX_HOURS,
  eipAllocationId: 'eipalloc-0d1bc6d85895a5410',
  defaultInstanceType: new InstanceType('c6g.large'),
  vpc: fleet.vpc,
})
// configure the expiration after 6 hours
fleet2.expireAfter(Duration.hours(6))

// print the instanceId from each spot fleet
new CfnOutput(stack, 'SpotFleetInstanceId', { value: fleet.instanceId })
new CfnOutput(stack, 'SpotFleet2InstanceId', { value: fleet2.instanceId })

```

# Create spot instances without duration block

```ts
const fleet = new SpotFleet(stack, 'SpotFleet', {
  blockDuration: BlockDuration.NONE,
})
```

NOTE: This kind of spot instance will be interrupted by AWS. However the fleet is using type [maintain](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet.html#spot-fleet-allocation-strategy), the fleet can be refulfilled.

# ARM64 and Graviton 2 support

`cdk-spot-one` selects the latest Amazon Linux 2 AMI for your `ARM64` instances. Simply select the instance types with the `defaultInstanceType` property and the `SpotFleet` will auto configure correct AMI for the instance.


```ts
defaultInstanceType: new InstanceType('c6g.large')
```

# SSH connect

By default the `cdk-spot-one` does not bind any SSH public key for you on the instance. You are encouraged to use `ec2-instance-connect` to send your public key from local followed by one-time SSH connect.

For example:

```sh
pubkey="$HOME/.ssh/aws_2020_id_rsa.pub"
echo "sending public key to ${instanceId}"
aws ec2-instance-connect send-ssh-public-key --instance-id ${instanceId} --instance-os-user ec2-user \
--ssh-public-key file://${pubkey} --availability-zone ${az} 
```
## npx ec2-connect INSTANCE_ID

To connect to the instance, run `npx ec2-connect` as below:

```sh
$ npx ec2-connect i-01f827ab9de7b93a9
```

or 

```sh
$ npx ec2-connect i-01f827ab9de7b93a9 ~/.ssh/other_public_key_path
```
If you are using different SSH public key(default is ~/.ssh/id_rsa.pub)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.6.161](https://github.com/pahud/cdk-spot-one/compare/v0.6.160...v0.6.161) (2021-04-07)


### Features

* support SpotInstance ([6c1f01c](https://github.com/pahud/cdk-spot-one/commit/6c1f01cb474a775c5a685ba7cc56eb232d01f0d4))

### [0.6.160](https://github.com/pahud/cdk-spot-one/compare/v0.6.159...v0.6.160) (2021-02-24)

### [0.6.159](https://github.com/pahud/cdk-spot-one/compare/v0.6.158...v0.6.159) (2021-02-23)

### [0.6.158](https://github.com/pahud/cdk-spot-one/compare/v0.6.157...v0.6.158) (2021-02-21)

### [0.6.157](https://github.com/pahud/cdk-spot-one/compare/v0.6.156...v0.6.157) (2021-02-20)

### [0.6.156](https://github.com/pahud/cdk-spot-one/compare/v0.6.155...v0.6.156) (2021-02-19)

### [0.6.155](https://github.com/pahud/cdk-spot-one/compare/v0.6.154...v0.6.155) (2021-02-18)

### [0.6.154](https://github.com/pahud/cdk-spot-one/compare/v0.6.153...v0.6.154) (2021-02-17)

### [0.6.153](https://github.com/pahud/cdk-spot-one/compare/v0.6.152...v0.6.153) (2021-02-16)

### [0.6.152](https://github.com/pahud/cdk-spot-one/compare/v0.6.151...v0.6.152) (2021-02-15)

### [0.6.151](https://github.com/pahud/cdk-spot-one/compare/v0.6.150...v0.6.151) (2021-02-14)

### [0.6.150](https://github.com/pahud/cdk-spot-one/compare/v0.6.149...v0.6.150) (2021-02-13)

### [0.6.149](https://github.com/pahud/cdk-spot-one/compare/v0.6.148...v0.6.149) (2021-02-12)

### [0.6.148](https://github.com/pahud/cdk-spot-one/compare/v0.6.147...v0.6.148) (2021-02-11)

### [0.6.147](https://github.com/pahud/cdk-spot-one/compare/v0.6.146...v0.6.147) (2021-02-10)


### Features

* support SpotInstance ([#199](https://github.com/pahud/cdk-spot-one/issues/199)) ([0c31dfb](https://github.com/pahud/cdk-spot-one/commit/0c31dfbfe99229b3bb782cc8f6d115133fa12f15)), closes [#193](https://github.com/pahud/cdk-spot-one/issues/193)

### [0.6.146](https://github.com/pahud/cdk-spot-one/compare/v0.6.145...v0.6.146) (2021-02-10)

### [0.6.145](https://github.com/pahud/cdk-spot-one/compare/v0.6.144...v0.6.145) (2021-02-09)

### [0.6.144](https://github.com/pahud/cdk-spot-one/compare/v0.6.143...v0.6.144) (2021-02-08)

### [0.6.143](https://github.com/pahud/cdk-spot-one/compare/v0.6.142...v0.6.143) (2021-02-07)

### [0.6.142](https://github.com/pahud/cdk-spot-one/compare/v0.6.141...v0.6.142) (2021-02-06)

### [0.6.141](https://github.com/pahud/cdk-spot-one/compare/v0.6.140...v0.6.141) (2021-02-05)

### [0.6.140](https://github.com/pahud/cdk-spot-one/compare/v0.6.139...v0.6.140) (2021-02-04)

### [0.6.139](https://github.com/pahud/cdk-spot-one/compare/v0.6.138...v0.6.139) (2021-02-04)


### Features

* support instanceInterruptionBehavior property ([2a51016](https://github.com/pahud/cdk-spot-one/commit/2a51016070b808cd408c72fe558db9f82975b709))
* support instanceInterruptionBehavior property ([#187](https://github.com/pahud/cdk-spot-one/issues/187)) ([ad036ad](https://github.com/pahud/cdk-spot-one/commit/ad036ad82851e0689317a96a7753af6dcd3f9041))


### Bug Fixes

* instanceInterruptionBehavior not honored in SpotFleet ([ba3d908](https://github.com/pahud/cdk-spot-one/commit/ba3d90888b3b23e1402497cba274a7e526d1bb46))
* lockfile ([a60fea6](https://github.com/pahud/cdk-spot-one/commit/a60fea673be94e5ec367912695db05cae3f3a63c))

### [0.6.138](https://github.com/pahud/cdk-spot-one/compare/v0.6.137...v0.6.138) (2021-02-03)

### [0.6.137](https://github.com/pahud/cdk-spot-one/compare/v0.0.1...v0.6.137) (2021-02-02)

### [0.6.33](https://github.com/pahud/cdk-spot-one/compare/v0.0.1...v0.6.33) (2021-02-02)

### [0.6.136](https://github.com/pahud/cdk-spot-one/compare/v0.6.135...v0.6.136) (2021-02-02)

### [0.6.135](https://github.com/pahud/cdk-spot-one/compare/v0.6.134...v0.6.135) (2021-02-01)

### [0.6.134](https://github.com/pahud/cdk-spot-one/compare/v0.6.133...v0.6.134) (2021-01-12)

### [0.6.133](https://github.com/pahud/cdk-spot-one/compare/v0.6.132...v0.6.133) (2021-01-11)

### [0.6.132](https://github.com/pahud/cdk-spot-one/compare/v0.6.131...v0.6.132) (2021-01-10)


src/test/index.test.ts
import { SpotFleet, InstanceInterruptionBehavior } from '../index';
import { App, Stack, Duration } from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import '@aws-cdk/assert/jest';
import { BlockDuration } from '../spot';

describe('Spot Fleet tests', () => {

  test('create the HTTP API', () => {
    const mockApp = new App();
    const stack = new Stack(mockApp, 'testing-stack');

    const fleet = new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.SIX_HOURS,
      instanceInterruptionBehavior: InstanceInterruptionBehavior.HIBERNATE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      eipAllocationId: 'eipalloc-0d1bc6d85895a5410',
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
    });
    // fleet to expire after 6 hours
    fleet.expireAfter(Duration.hours(6));
    fleet.defaultSecurityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(80));
    expect(stack).toHaveResource('AWS::EC2::SpotFleet');
    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        InstanceMarketOptions: {
          MarketType: 'spot',
          SpotOptions: {
            BlockDurationMinutes: 360,
            InstanceInterruptionBehavior: 'hibernate',
          },
        },
      },
    });
    expect(stack).toHaveResource('AWS::EC2::SecurityGroup', {
      SecurityGroupIngress: [
        {
          CidrIp: '0.0.0.0/0',
          Description: 'from 0.0.0.0/0:22',
          FromPort: 22,
          IpProtocol: 'tcp',
          ToPort: 22,
        },
        {
          CidrIp: '0.0.0.0/0',
          Description: 'from 0.0.0.0/0:80',
          FromPort: 80,
          IpProtocol: 'tcp',
          ToPort: 80,
        },
      ],
    });
  });
  
  test('support additional user data', () => {
    const mockApp = new App();
    const stack = new Stack(mockApp, 'testing-stack');

    new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.SIX_HOURS,
      instanceInterruptionBehavior: InstanceInterruptionBehavior.HIBERNATE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
      additionalUserData: [ 
        'mycommand1',
        'mycommand2 arg1',
      ],
    });
  
    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        UserData: {
          'Fn::Base64': '#!/bin/bash\nyum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm\nyum install -y docker\nusermod -aG docker ec2-user\nusermod -aG docker ssm-user\nservice docker start\nmycommand1\nmycommand2 arg1',
        },
      },
    });
  });

  test('long time spot fleet', () => {
    const mockApp = new App();
    const stack = new Stack(mockApp, 'testing-stack');

    new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.NONE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
    });
  
    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        InstanceMarketOptions: {
          MarketType: 'spot',
          SpotOptions: {
            InstanceInterruptionBehavior: 'terminate',
          },
        },
      },
    });
  });
});


src/spot.ts
import * as path from 'path';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import { Construct, PhysicalName, Stack, Fn, CfnOutput, Duration, Lazy, CustomResource, Token } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';

const DEFAULT_INSTANCE_TYPE = 't3.large';

export class VpcProvider extends Stack {
  public static getOrCreate(scope: Construct) {
    const stack = Stack.of(scope);
    return stack.node.tryGetContext('use_default_vpc') === '1' ?
      ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true }) :
      stack.node.tryGetContext('use_vpc_id') ?
        ec2.Vpc.fromLookup(stack, 'Vpc', { vpcId: stack.node.tryGetContext('use_vpc_id') }) :
        new ec2.Vpc(stack, 'Vpc', { maxAzs: 3, natGateways: 1 });
  }
}

export enum BlockDuration {
  ONE_HOUR = 60,
  TWO_HOURS = 120,
  THREE_HOURS = 180,
  FOUR_HOURS = 240,
  FIVE_HOURS = 300,
  SIX_HOURS = 360,
  NONE = 0,
}

/**
 * Whether the worker nodes should support GPU or just standard instances
 */
export enum NodeType {
  /**
   * Standard instances
   */
  STANDARD = 'Standard',

  /**
   * GPU instances
   */
  GPU = 'GPU',

  /**
   * Inferentia instances
   */
  INFERENTIA = 'INFERENTIA',

  /**
   * ARM instances
   */
  ARM = 'ARM',


}

export enum InstanceInterruptionBehavior {
  HIBERNATE = 'hibernate',
  STOP = 'stop',
  TERMINATE = 'terminate'
}

export interface SpotFleetLaunchTemplateConfig {
  readonly spotfleet: SpotFleet;
  readonly launchTemplate: ILaunchtemplate;
}

export interface ILaunchtemplate {
  bind(spotfleet: SpotFleet): SpotFleetLaunchTemplateConfig;
}

export class LaunchTemplate implements ILaunchtemplate {
  public bind(spotfleet: SpotFleet): SpotFleetLaunchTemplateConfig {
    return {
      spotfleet,
      launchTemplate: this,
    };
  }
}

export interface SpotOneProps {
  /**
   * VPC for the spot fleet
   *
   * @default - new VPC will be created
   *
   */
  readonly vpc?: ec2.IVpc;

  /**
   * default EC2 instance type
   *
   * @default - t3.large
   */
  readonly defaultInstanceType?: ec2.InstanceType;

  /**
   * The behavior when a Spot Instance is interrupted
   *
   * @default - InstanceInterruptionBehavior.TERMINATE
   */
  readonly instanceInterruptionBehavior?: InstanceInterruptionBehavior;

  /**
   * IAM role for the spot instance
   */
  readonly instanceRole?: iam.IRole;

  /**
   * number of the target capacity
   *
   * @default - 1
   */
  readonly targetCapacity?: number;

  /**
   * custom AMI ID
   *
   * @default - The latest Amaozn Linux 2 AMI ID
   */
  readonly customAmiId?: string;

  /**
   * VPC subnet for the spot fleet
   *
   * @default - public subnet
   */
  readonly vpcSubnet?: ec2.SubnetSelection;

  /**
   * Security group for the spot fleet
   *
   * @default - allows TCP 22 SSH ingress rule
   */
  readonly securityGroup?: ec2.SecurityGroup;

  /**
   * SSH key name
   *
   * @default - no ssh key will be assigned
   */
  readonly keyName?: string;

  /**
   * Additional commands for user data
   *
   * @default - no additional user data
   */
  readonly additionalUserData?: string[];

  /**
   * Allocation ID for your existing Elastic IP Address.
   *
   * @defalt new EIP and its association will be created for the first instance in this spot fleet
   */
  readonly eipAllocationId?: string;

  /**
   * Auto assign a new EIP on this instance if `eipAllocationId` is not defined
   *
   * @default true
   */
  readonly assignEip?: boolean;

  /**
   * default EBS volume size for the spot instance
   *
   * @default 60;
   */
  readonly ebsVolumeSize?: number;

  /**
   * instance profile for the resource
   *
   * @default - create a new one
   */
  readonly instanceProfile?: iam.CfnInstanceProfile;
}

export abstract class SpotOne extends Construct {
  readonly vpc: ec2.IVpc;
  /**
   * The default security group of the instance, which only allows TCP 22 SSH ingress rule.
   */
  readonly defaultSecurityGroup: ec2.ISecurityGroup;
  readonly instanceId?: string;
  readonly instanceType?: string;
  readonly defaultInstanceType: ec2.InstanceType;
  readonly imageId: string;
  readonly userData: ec2.UserData;
  protected instanceRole?: iam.IRole;
  protected instanceProfile?: iam.CfnInstanceProfile;

  constructor(scope: Construct, id: string, props: SpotOneProps) {
    super(scope, id);

    this.vpc = props.vpc ?? VpcProvider.getOrCreate(this);
    this.defaultSecurityGroup = props.securityGroup || this.createSecurityGroup();
    this.instanceProfile = props.instanceProfile;
    this.defaultInstanceType = props.defaultInstanceType ?? new ec2.InstanceType(DEFAULT_INSTANCE_TYPE);

    this.imageId = props.customAmiId ??
      ec2.MachineImage.latestAmazonLinux({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
        cpuType: nodeTypeForInstanceType(this.defaultInstanceType) === NodeType.ARM ? ec2.AmazonLinuxCpuType.ARM_64 : undefined,
      }).getImage(this).imageId;

    this.userData = ec2.UserData.forLinux();

    /**
     * If not using custom AMI, we use amazon linux 2 and install the SSM agent and enable docker by default.
     * Otherwise, we simply do nothing here.
     */
    if (props.customAmiId == undefined) {
      this.userData.addCommands(
        'yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm',
        'yum install -y docker',
        'usermod -aG docker ec2-user',
        'usermod -aG docker ssm-user',
        'service docker start',
      );
    }

    if (props.additionalUserData) { this.userData.addCommands(...props.additionalUserData); }

    this.associateEip(props);
  }
  protected createInstanceRole(): iam.IRole {
    this.instanceRole = new iam.Role(this, 'InstanceRole', {
      roleName: PhysicalName.GENERATE_IF_NEEDED,
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
    });
    this.instanceRole.addManagedPolicy({
      managedPolicyArn: 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
    });
    return this.instanceRole;
  }
  protected createSecurityGroup(): ec2.SecurityGroup {
    const securityGroup = new ec2.SecurityGroup(this, 'SpotFleetSg', {
      vpc: this.vpc,
    });
    securityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(22));
    return securityGroup;
  }
  protected associateEip(props: SpotOneProps) {
    // EIP association
    if (props.eipAllocationId) {
      new ec2.CfnEIPAssociation(this, 'EipAssocation', {
        allocationId: props.eipAllocationId,
        instanceId: this.instanceId,
      });
    } else if (props.assignEip !== false) {
      new ec2.CfnEIP(this, 'EIP', {
        instanceId: this.instanceId,
      });
    }
  }
  protected createInstanceProfile(role: iam.IRole): iam.CfnInstanceProfile {
    return new iam.CfnInstanceProfile(this, 'InstanceProfile', {
      roles: [role.roleName],
    });
  }
}


export interface SpotInstanceProps extends SpotOneProps {}

export class SpotInstance extends SpotOne {
  readonly instanceId?: string;
  readonly instanceType?: string;
  constructor(scope: Construct, id: string, props: SpotInstanceProps = {}) {
    super(scope, id, props);

    const spotInstance = new ec2.Instance(this, 'SpotInstance', {
      vpc: this.vpc,
      instanceType: this.defaultInstanceType,
      machineImage: ec2.MachineImage.latestAmazonLinux({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
        cpuType: nodeTypeForInstanceType(this.defaultInstanceType) === NodeType.ARM ? ec2.AmazonLinuxCpuType.ARM_64 : undefined,
      }),
      keyName: props.keyName,
      securityGroup: this.defaultSecurityGroup,
      role: this.instanceRole,
      userData: this.userData,
      vpcSubnets: props.vpcSubnet ?? {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      blockDevices: [
        {
          deviceName: '/dev/xvda',
          volume: ec2.BlockDeviceVolume.ebs(props.ebsVolumeSize ?? 60),
        },
      ],
    });
    const cfnInstance = spotInstance.node.defaultChild as ec2.CfnInstance;
    // create custom launch template reousrce
    const launchTemplate = new LaunchTemplateResource(this, 'launchTemplateForInstance', {
      instanceMarketOptions: {
        marketType: 'spot',
        spotOptions: {
          instanceInterruptionBehavior: props.instanceInterruptionBehavior,
          spotInstanceType: props.instanceInterruptionBehavior === InstanceInterruptionBehavior.TERMINATE ? 'one-time'
            : 'persistent',
        },
      },
      iamInstanceProfile: this.instanceProfile,
    });
    cfnInstance.addPropertyOverride('LaunchTemplate', {
      LaunchTemplateId: launchTemplate.resource.ref,
      Version: launchTemplate.resource.attrLatestVersionNumber,
    });
    // As we can't specify IamInstanceProfile both in launch template and instane property
    // we need delete the property here.
    cfnInstance.addPropertyDeletionOverride('IamInstanceProfile');

    this.instanceId = spotInstance.instanceId;
    this.instanceType = this.defaultInstanceType.toString();
    new CfnOutput(this, 'PublicIpAddress', { value: cfnInstance.attrPublicIp });
  }
}

export interface BaseSpotFleetProps extends SpotOneProps {
  /**
   * reservce the spot instance as spot block with defined duration
   *
   * @default - BlockDuration.ONE_HOUR
   */
  readonly blockDuration?: BlockDuration;

  /**
   * the time when the spot fleet allocation starts
   *
   * @default - no expiration
   */
  readonly validFrom?: string;

  /**
   * the time when the spot fleet allocation expires
   *
   * @default - no expiration
   */
  readonly validUntil?: string;

  /**
   * terminate the instance when the allocation is expired
   *
   * @default - true
   */
  readonly terminateInstancesWithExpiration?: boolean;

  /**
   * blockDeviceMappings for config instance.
   *
   * @default - from ami config.
   */
  readonly blockDeviceMappings?: ec2.CfnLaunchTemplate.BlockDeviceMappingProperty[] | undefined;
}

export interface SpotFleetProps extends BaseSpotFleetProps {
  /**
   * Launch template for the spot fleet
   */
  readonly launchTemplate?: ILaunchtemplate;

  /**
   * Whether to create spot instance only instead of a fleet.
   *
   * @default false;
   */
  readonly instanceOnly?: boolean;
}

export class SpotFleet extends SpotOne {
  // readonly instanceRole: iam.IRole;
  // readonly defaultInstanceType: ec2.InstanceType;
  readonly targetCapacity?: number;
  readonly spotFleetId: string;
  readonly launchTemplate: ILaunchtemplate;
  // readonly vpc: ec2.IVpc;
  /**
   * the first instance id in this fleet
   */
  readonly instanceId?: string;
  /**
   * instance type of the first instance in this fleet
   */
  readonly instanceType?: string;
  /**
   * SpotFleetRequestId for this spot fleet
   */
  readonly spotFleetRequestId?: string;
  /**
   * The behavior when a Spot Instance is interrupted
   *
   * @default terminate
   */
  readonly instanceInterruptionBehavior?: InstanceInterruptionBehavior;
  /**
   * The time when the the fleet allocation will expire
   */
  private validUntil?: string;


  constructor(scope: Construct, id: string, props: SpotFleetProps = {}) {
    super(scope, id, props);
    this.spotFleetId = id;
    this.launchTemplate = props.launchTemplate ?? new LaunchTemplate();
    this.targetCapacity = props.targetCapacity ?? 1;
    this.validUntil = props.validUntil;

    const stack = Stack.of(this);

    this.instanceProfile = props.instanceProfile ?? props.instanceRole ? this.createInstanceProfile(props.instanceRole!)
      : this.createInstanceProfile(this.createInstanceRole());

    const lt = new LaunchTemplateResource(this, 'LaunchTemplate', {
      blockDeviceMappings: props.blockDeviceMappings,
      defaultInstanceType: this.defaultInstanceType,
      iamInstanceProfile: this.instanceProfile,
      imageId: this.imageId,
      instanceMarketOptions: {
        marketType: 'spot',
        spotOptions: {
          blockDurationMinutes: (props.blockDuration == BlockDuration.NONE) ? undefined : (props.blockDuration ?? BlockDuration.ONE_HOUR),
          instanceInterruptionBehavior: props.instanceInterruptionBehavior ?? InstanceInterruptionBehavior.TERMINATE,
        },
      },
      keyName: props.keyName,
      securityGroup: this.defaultSecurityGroup.connections.securityGroups,
      userData: this.userData,
    });

    const spotFleetRole = new iam.Role(this, 'FleetRole', {
      assumedBy: new iam.ServicePrincipal('spotfleet.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonEC2SpotFleetTaggingRole'),
      ],
    });

    const vpcSubnetSelection = props.vpcSubnet ?? {
      subnetType: ec2.SubnetType.PUBLIC,
    };
    const subnetConfig = this.vpc.selectSubnets(vpcSubnetSelection).subnets.map(s => ({
      subnetId: s.subnetId,
    }));

    const cfnSpotFleet = new ec2.CfnSpotFleet(this, id, {
      spotFleetRequestConfigData: {
        launchTemplateConfigs: [
          {
            launchTemplateSpecification: {
              launchTemplateId: lt.resource.ref,
              version: lt.resource.attrLatestVersionNumber,
            },
            overrides: subnetConfig,
          },
        ],
        iamFleetRole: spotFleetRole.roleArn,
        targetCapacity: props.targetCapacity ?? 1,
        validFrom: props.validFrom,
        validUntil: Lazy.string({ produce: () => this.validUntil }),
        terminateInstancesWithExpiration: props.terminateInstancesWithExpiration ?? true,
        instanceInterruptionBehavior: props.instanceInterruptionBehavior ?? InstanceInterruptionBehavior.TERMINATE,
      },
    });
    new CfnOutput(stack, 'SpotFleetId', { value: cfnSpotFleet.ref });
    const onEvent = new lambda.Function(this, 'OnEvent', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../eip-handler')),
      handler: 'index.on_event',
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(60),
    });

    const isComplete = new lambda.Function(this, 'IsComplete', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../eip-handler')),
      handler: 'index.is_complete',
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(60),
      role: onEvent.role,
    });

    const myProvider = new cr.Provider(this, 'MyProvider', {
      onEventHandler: onEvent,
      isCompleteHandler: isComplete, // optional async "waiter"
      logRetention: logs.RetentionDays.ONE_DAY, // default is INFINITE
    });

    onEvent.addToRolePolicy(new iam.PolicyStatement({
      actions: ['ec2:DescribeSpotFleetInstances'],
      resources: ['*'],
    }));

    const fleetInstances = new CustomResource(this, 'SpotFleetInstances', {
      serviceToken: myProvider.serviceToken,
      properties: {
        SpotFleetRequestId: cfnSpotFleet.ref,
      },
    });

    fleetInstances.node.addDependency(cfnSpotFleet);

    this.instanceId = Token.asString(fleetInstances.getAtt('InstanceId'));
    this.instanceType = Token.asString(fleetInstances.getAtt('InstanceType'));
    this.spotFleetRequestId = Token.asString(fleetInstances.getAtt('SpotInstanceRequestId'));
    new CfnOutput(stack, 'InstanceId', { value: this.instanceId });
  }
  public expireAfter(duration: Duration) {
    const date = new Date();
    date.setSeconds(date.getSeconds() + duration.toSeconds());
    this.validUntil = date.toISOString();
  }
}

export interface LaunchTemplateProps {
  readonly imageId?: string;
  readonly defaultInstanceType?: ec2.InstanceType;
  readonly keyName?: string;
  readonly userData?: ec2.UserData;
  /**
   * blockDeviceMappings for config instance.
   *
   * @default - from ami config.
   */
  readonly blockDeviceMappings?: ec2.CfnLaunchTemplate.BlockDeviceMappingProperty[] | undefined;
  readonly instanceMarketOptions?: ec2.CfnLaunchTemplate.InstanceMarketOptionsProperty;
  readonly securityGroup?: ec2.ISecurityGroup[];
  readonly iamInstanceProfile?: iam.CfnInstanceProfile;
}

export class LaunchTemplateResource extends Construct {
  readonly defaultInstanceType: ec2.InstanceType;
  readonly resource: ec2.CfnLaunchTemplate;
  constructor(scope: Construct, id: string, props: LaunchTemplateProps = {}) {
    super(scope, id);

    this.defaultInstanceType = props.defaultInstanceType ?? new ec2.InstanceType(DEFAULT_INSTANCE_TYPE);
    const imageId = props.imageId ??
      ec2.MachineImage.latestAmazonLinux({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
        cpuType: nodeTypeForInstanceType(this.defaultInstanceType) === NodeType.ARM ? ec2.AmazonLinuxCpuType.ARM_64 : undefined,
      }).getImage(this).imageId;

    const userData = props.userData ?? ec2.UserData.forLinux();

    this.resource = new ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        imageId,
        instanceType: this.defaultInstanceType.toString(),
        userData: Fn.base64(userData.render()),
        keyName: props.keyName,
        tagSpecifications: [
          {
            resourceType: 'instance',
            tags: [
              {
                key: 'Name',
                value: `${Stack.of(this).stackName}/${id}`,
              },
            ],
          },
        ],
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        instanceMarketOptions: props.instanceMarketOptions,
        securityGroupIds: props.securityGroup?.map(s => s.securityGroupId),
        iamInstanceProfile: props.iamInstanceProfile ? {
          arn: props.iamInstanceProfile.attrArn,
        } : undefined,
      },
    });
  }
}

const GRAVITON_INSTANCETYPES = ['a1'];
const GRAVITON2_INSTANCETYPES = ['c6g', 'm6g', 'r6g'];
const GPU_INSTANCETYPES = ['p2', 'p3', 'g4'];
const INFERENTIA_INSTANCETYPES = ['inf1'];

function nodeTypeForInstanceType(instanceType: ec2.InstanceType) {
  return GPU_INSTANCETYPES.includes(instanceType.toString().substring(0, 2)) ? NodeType.GPU :
    INFERENTIA_INSTANCETYPES.includes(instanceType.toString().substring(0, 4)) ? NodeType.INFERENTIA :
      GRAVITON2_INSTANCETYPES.includes(instanceType.toString().substring(0, 3)) ? NodeType.ARM :
        GRAVITON_INSTANCETYPES.includes(instanceType.toString().substring(0, 2)) ? NodeType.ARM :
          NodeType.STANDARD;
}
